File: Zig/src/materials.zig
const std = @import("std");
const rand = std.rand;
const math = std.math;

const Viewport = @import("viewport.zig").Viewport;
const Scene = @import("viewport.zig").Scene;
const Sphere = @import("object.zig").Sphere;
const Hit = @import("object.zig").Hit;
const NO_HIT = @import("object.zig").NO_HIT;
const Ray = @import("ray.zig").Ray;
const unit_vec = @import("ray.zig").unit_vec;
const Vec3 = @Vector(3, f32);
const vec = @import("ray.zig");

pub fn rand_vec_unit() Vec3 {
    var prng = std.rand.DefaultPrng.init(@truncate(u64, @bitCast(u128, std.time.nanoTimestamp())));
    return unit_vec(blk: while (true) {
        var x = Vec3{ prng.random().float(f32), prng.random().float(f32), prng.random().float(f32) } * Vec3{ 2.0, 2.0, 2.0 } - Vec3{ 1.0, 1.0, 1.0 };
        if (x[0] * x[0] + x[1] * x[1] + x[2] * x[2] <= 1.0) {
            break :blk x;
        }
    });
}

fn diffuse(h: Hit, _: Ray) Ray {
    // println!("diff");
    var target = h.normal + rand_vec_unit();
    return Ray{ .origin = h.point, .direction = target };
}
fn metallic(h: Hit, r: Ray) Ray {
    return Ray{ .origin = h.point, .direction = vec.reflect(vec.unit_vec(r.direction), h.normal) };
}
fn metal_fuzzy03(h: Hit, r: Ray) Ray {
    return Ray{ .origin = h.point, .direction = vec.reflect(vec.unit_vec(r.direction), h.normal) + rand_vec_unit() * Vec3{ 0.3, 0.3, 0.3 } };
}
fn refract(uv: Vec3, n: Vec3, etai_over_etat: f32) Vec3 {
    var cos_theta = vec.dot_product(-uv, n);
    if (cos_theta > 1.0) {
        cos_theta = 1.0;
    }
    var r_out_perp = (uv + n * @splat(3, cos_theta)) * @splat(3, etai_over_etat);
    var r_out_parallel = n * @splat(3, -math.sqrt(math.fabs(1.0 - vec.vec_len2(r_out_perp))));
    return r_out_perp + r_out_parallel;
}

fn reflectance(cosine: f32, ref_idx: f32) f32 {
    // Use Schlick's approximation for reflectance.
    var r0 = (1.0 - ref_idx) / (1.0 + ref_idx);
    r0 = r0 * r0;
    return r0 + (math.pow(f32, (1.0 - r0) * (1.0 - cosine), 5));
}

pub const Material = struct {
    metallicness: f32 = 0,
    opacity: f32 = 0,
    ir: f32 = 1.0,
    fn init(metallicness: f32, opacity: f32) Material {
        return Material{ .metallicness = metallicness, .opacity = opacity };
    }

    fn onHit(self: *Material, h: Hit, r: Ray) Ray {
        var prng = std.rand.DefaultPrng.init(@truncate(u64, @bitCast(u128, std.time.nanoTimestamp())));
        if (self.opacity > 0.0) {
            var n = h.normal;
            var front_face = true;
            if (vec.dot_product(r.direction, h.normal) > 0.0) {
                n = -h.normal;
                front_face = false;
            }
            var refraction_ratio = if (front_face) 1.0 / self.ir else self.ir;

            var unit_direction = vec.unit_vec(r.direction);
            var cos_theta = vec.dot_product(-unit_direction, n);
            if (cos_theta > 1.0) {
                cos_theta = 1.0;
            }
            var sin_theta = math.sqrt(1.0 - cos_theta * cos_theta);

            var cannot_refract = refraction_ratio * sin_theta > 1.0;
            var reflectancev = reflectance(cos_theta, refraction_ratio);
            // eprintln!("ff: {} can refract: {} ref_ratio: {}", front_face, !cannot_refract, refraction_ratio);
            var direction = if (cannot_refract or reflectancev > prng.random().float(f32))
                // eprintln!("reflect");
                vec.reflect(unit_direction, n)
            else
                // eprintln!("ud: {:?} hn: {:?}", unit_direction, n);
                refract(unit_direction, n, refraction_ratio);

            return Ray{ .origin = h.point, .direction = direction };
        }
        var sc = diffuse(h, r).direction * @splat(3, 1.0 - self.metallicness);
        var reflect = metallic(h, r);
        reflect.direction = reflect.direction * @splat(3, self.metallicness) + sc;
        return reflect;
    }
};

pub const diffuseM = Material{};
pub const metallicM = Material{ .metallicness = 1.0 };
pub const glassM = Material{ .metallicness = 1.0, .opacity = 1.0, .ir = 1.5 };
pub const glassRM = Material{ .metallicness = 1.0, .opacity = 1.0, .ir = 1 / glassM.ir };
pub const fuzzy3M = Material{ .metallicness = 0.7 };

fn ray_color(r: Ray, scene: Scene, depth: usize) Vec3 {
    if (depth <= 0) {
        return Vec3{ 0.0, 0.0, 0.0 };
    }
    const mint = 0.001;
    const maxt = 1000;
    var min_hit = NO_HIT;
    // var changed = false;
    for (scene.spheres.items) |s| {
        var hit = s.collision(r, mint, maxt);
        if (hit.equal(&NO_HIT)) continue;
        if (min_hit.equal(&NO_HIT) or hit.t < min_hit.t) {
            min_hit = hit;
            // changed = true;
        }
    }

    if (min_hit.equal(&NO_HIT)) {
        var unit_direction = unit_vec(r.direction);
        var t = 0.5 * (unit_direction[1] + 1.0);
        return Vec3{ (1.0 - t) + t * 0.5, ((1.0 - t) + t * 0.7), 1.0 };
    } else {
        return ray_color(min_hit.mat.onHit(min_hit, r), scene, depth - 1) * min_hit.col;
    }
}

pub fn diffuseTest() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};

    const w = 900;
    const h = 600;
    const samples = 100;
    var view = Viewport{ .width = w, .height = h, .aspect_ratio = @intToFloat(f32, w) / @intToFloat(f32, h), .samples = samples };
    var spheres = try std.ArrayList(Sphere).initCapacity(gpa.allocator(), 4);
    try spheres.append(Sphere{ .origin = Vec3{ 0.5, 0.0, -1.0 }, .radius = 0.5, .mat = diffuseM, .col = Vec3{ 1.0, 0.2, 0.2 } });
    try spheres.append(Sphere{ .origin = Vec3{ -0.5, 0.0, -1.0 }, .radius = 0.5, .mat = diffuseM, .col = Vec3{ 0.8, 0.8, 1.0 } });
    try spheres.append(Sphere{ .origin = Vec3{ 0.0, 0.2, -2.0 }, .radius = 1.0, .mat = diffuseM, .col = Vec3{ 0.2, 1.0, 0.2 } });
    try spheres.append(Sphere{ .origin = Vec3{ 0.0, 0.0, 0.0 }, .radius = 1.0 });

    var scene = Scene{ .spheres = spheres };
    defer scene.deinit();

    try view.Render(ray_color, scene, "diffuse_test.png");
}

pub fn metalTest() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};

    const w = 800;
    const h = 600;
    const samples = 100;
    var view = Viewport{ .width = w, .height = h, .aspect_ratio = @intToFloat(f32, w) / @intToFloat(f32, h), .samples = samples, .depth = 10 };
    var spheres = try std.ArrayList(Sphere).initCapacity(gpa.allocator(), 4);
    try spheres.append(Sphere{ .origin = Vec3{ -0.5, 0.0, -1.0 }, .radius = 0.5, .mat = fuzzy3M, .col = Vec3{ 0.6, 0.6, 0.6 } });
    try spheres.append(Sphere{ .origin = Vec3{ 0.5, 0.0, -1.0 }, .radius = 0.5, .mat = metallicM, .col = Vec3{ 0.5, 0.9, 0.9 } });
    try spheres.append(Sphere{ .origin = Vec3{ 0.0, 0.0, -2.0 }, .radius = 1.0, .mat = diffuseM, .col = Vec3{ 0.5, 1.0, 0.0 } });
    try spheres.append(Sphere{ .origin = Vec3{ 0.0, -1000.9, 0.0 }, .radius = 1000.0, .mat = diffuseM, .col = Vec3{ 1.0, 0.0, 1.0 } });

    var scene = Scene{ .spheres = spheres };
    defer scene.deinit();

    try view.Render(ray_color, scene, "metal_test.png");
}

pub fn glassTest() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};

    const w = 800;
    const h = 600;
    const samples = 100;
    var view = Viewport{ .width = w, .height = h, .aspect_ratio = @intToFloat(f32, w) / @intToFloat(f32, h), .samples = samples, .depth = 10 };
    var spheres = try std.ArrayList(Sphere).initCapacity(gpa.allocator(), 4);
    // try spheres.append(Sphere{ .origin = Vec3{ -0.5, 0.0, -1.0 }, .radius = 0.5, .mat = fuzzy3M, .col = Vec3{ 0.6, 0.6, 0.6 } });
    // try spheres.append(Sphere{ .origin = Vec3{ 0.5, 0.0, -1.0 }, .radius = 0.5, .mat = metallicM, .col = Vec3{ 0.5, 0.9, 0.9 } });
    try spheres.append(Sphere{ .origin = Vec3{ 0.0, 0.0, -1.0 }, .radius = 0.5, .mat = glassM, .col = Vec3{ 1.0, 1.0, 1.0 } });
    try spheres.append(Sphere{ .origin = Vec3{ 0.0, 0.0, -1.0 }, .radius = 0.35, .mat = glassRM, .col = Vec3{ 1.0, 1.0, 1.0 } });
    try spheres.append(Sphere{ .origin = Vec3{ 0.0, -100.5, -1.0 }, .radius = 100.0, .mat = diffuseM, .col = Vec3{ 1.0, 0.0, 1.0 } });

    var scene = Scene{ .spheres = spheres };
    defer scene.deinit();

    try view.Render(ray_color, scene, "glass_test.png");
}

// Sphere::new(Vec3 {x: -0.5, y: 0.0, z: -1.0,}, 0.5, Some(Vec3::new(0.6, 0.6, 0.6)), Some(&diffuseM)),
// Sphere::new(Vec3 {x: 0.5, y: 0.0, z: -1.0,}, 0.5, Some(Vec3::new(0.5, 0.9, 0.9)), Some(&metal)),
// Sphere::new(Vec3 {x: 0.0, y: 0.0, z: -2.0,}, 1.0, Some(Vec3::new(0.5, 1.0, 0.0)), Some(&diffuseM)),
// Sphere::new(Vec3 {x: 0.0, y: 0.0, z: 0.0,}, 1.0, Some(Vec3::new(0.8, 0.5, 1.0)), Some(&empty))

File: Rust/src/objects/materials.rs
use super::{Hit, JsonValue};
use crate::vec3::{ray::Ray, vec3::Vec3};
use rand::random;

pub fn lambertian_scatter_pdf(cos_theta: f32) -> f32 {
    // let cos_theta = Vec3::dot(&hit.normal, r_out.direction.unit());
    // let cos_theta = r_out.direction.length() * 0.5;
    return if cos_theta > 0.0 {
        cos_theta * std::f32::consts::FRAC_1_PI
    } else {
        0.0
    };
}
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Material {
    pub metallicness: f32,
    pub opacity: f32,
    pub ir: f32,
    pub emmited: Vec3,
}
impl Into<JsonValue> for Material {
    fn into(self) -> JsonValue {
        json::object! {
            metallicness: self.metallicness,
            opacity: self.opacity,
            ir: self.ir
        }
    }
}
impl TryFrom<JsonValue> for Material {
    type Error = crate::viewport::errors::ParseError;

    fn try_from(value: JsonValue) -> Result<Self, Self::Error> {
        Ok(Material {
            metallicness: match value["metallicness"].as_f32() {
                Some(x) => x,
                None => return Err(Self::Error { source: None }),
            },
            opacity: match value["opacity"].as_f32() {
                Some(x) => x,
                None => return Err(Self::Error { source: None }),
            },
            ir: match value["ir"].as_f32() {
                Some(x) => x,
                None => return Err(Self::Error { source: None }),
            },
            emmited: Vec3 {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            },
        })
    }
}
impl Material {
    pub fn new(metallicness: f32, opacity: f32, ir: f32) -> Self {
        Self {
            metallicness: metallicness,
            opacity: opacity,
            ir: ir,
            emmited: Vec3 {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            },
        }
    }
    pub fn new_m(metallicness: f32) -> Self {
        Self {
            metallicness: metallicness,
            opacity: 0.0,
            ir: 1.0,
            emmited: Vec3 {
                x: 0.0,
                y: 0.0,
                z: 0.0,
            },
        }
    }
    pub fn new_emmiting(metallicness: f32, opacity: f32, ir: f32, emmited: Vec3) -> Self {
        Self {
            metallicness: metallicness,
            opacity,
            ir,
            emmited,
        }
    }

    fn refract(uv: Vec3, n: Vec3, etai_over_etat: f32) -> Vec3 {
        let mut cos_theta = (-uv).dot(n);
        if cos_theta > 1.0 {
            cos_theta = 1.0
        }
        let r_out_perp = (uv + n * cos_theta) * etai_over_etat;
        let r_out_parallel = n * -(1.0 - r_out_perp.length2()).abs().sqrt();
        return r_out_perp + r_out_parallel;
    }
    fn reflectance(cosine: f32, ref_idx: f32) -> f32 {
        // Use Schlick's approximation for reflectance.
        let mut r0 = (1.0 - ref_idx) / (1.0 + ref_idx);
        r0 = r0 * r0;
        return r0 + (1.0 - r0) * (1.0 - cosine).powi(5);
    }

    pub fn on_hit(&self, h: Hit, r: Ray) -> (Ray, Vec3, f32) {
        if self.opacity > 0.0 {
            let n;
            let front_face = if r.direction.dot(h.normal) > 0.0 {
                n = -h.normal;
                false
            } else {
                n = h.normal;
                true
            };
            let refraction_ratio = if front_face { 1.0 / self.ir } else { self.ir };

            let unit_direction = r.direction.unit();
            let mut cos_theta = (-unit_direction).dot(n);
            if cos_theta > 1.0 {
                cos_theta = 1.0;
            }
            let sin_theta = (1.0 - cos_theta * cos_theta).sqrt();

            let cannot_refract = refraction_ratio * sin_theta > 1.0;
            let reflectance = Self::reflectance(cos_theta, refraction_ratio);
            // eprintln!("ff: {} can refract: {} ref_ratio: {}", front_face, !cannot_refract, refraction_ratio);
            let direction = if cannot_refract || reflectance > random::<f32>() {
                // eprintln!("reflect");
                unit_direction.reflect(n)
            } else {
                // eprintln!("ud: {:?} hn: {:?}", unit_direction, n);
                Self::refract(unit_direction, n, refraction_ratio)
            };

            return (
                Ray::new_with_time(h.point, direction, r.time),
                self.emmited,
                0.0,
            );
        }
        // eprintln!("reflect");
        let sc = diffuse(h, r).direction;
        let mut reflect = Ray::new_with_time(h.point, r.direction.unit().reflect(h.normal), r.time);
        reflect.direction = reflect.direction * self.metallicness + sc * (1.0 - self.metallicness);
        return (
            reflect,
            self.emmited,
            if self.metallicness != 1.0 {
                sc.dot(h.normal)
            } else {
                0.0
            },
        );
    }
}

pub const METALLIC_M: Material = Material {
    metallicness: 1.0,
    opacity: 0.0,
    ir: 1.0,
    emmited: Vec3 {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    },
};

pub const SCATTER_M: Material = Material {
    metallicness: 0.0,
    opacity: 0.0,
    ir: 1.0,
    emmited: Vec3 {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    },
};

pub const FUZZY3_M: Material = Material {
    metallicness: 0.7,
    opacity: 0.0,
    ir: 1.0,
    emmited: Vec3 {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    },
};

pub const GLASS_M: Material = Material {
    metallicness: 1.0,
    opacity: 1.0,
    ir: 1.50,
    emmited: Vec3 {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    },
};

pub const GLASSR_M: Material = Material {
    metallicness: 1.0,
    opacity: 1.0,
    ir: 1.0 / GLASS_M.ir,
    emmited: Vec3 {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    },
};

pub const EMPTY_M: Material = SCATTER_M;
fn diffuse(hit: Hit, r: Ray) -> Ray {
    // println!("diff");
    let target = hit.normal + Vec3::random_unit_vec();
    if target.close_to_zero() {
        return Ray {
            origin: hit.point,
            direction: hit.normal,
            time: r.time,
        };
    }
    return Ray {
        origin: hit.point,
        direction: target,
        time: r.time,
    };
}
